### PR TITLE
Ensure reg.sock exists while the registrar is running

### LINKED ISSUE #57: Ensure reg.sock exists while the registrar is running
While the registrar is running, the kubelet registration socket it installs could be removed erroneously by a different process (for example if two instances of the same driver are running in the cluster). The registrar should continuously ensure that the socket exists during its lifetime (and subsequently removed before it terminates).

/cc @davidz627 @msau42

### PR DESCRIPTION
**Pull request by zhucan**
Signed-off-by: zhuc <zhucan.k8s@gmail.com>

**What type of PR is this?**
> /kind bug

**What this PR does / why we need it**:
Ensure reg.sock exists while the registrar is running

**Which issue(s) this PR fixes**:
Fixes #57 

**Special notes for your reviewer**:
@jingxu97 

**Does this PR introduce a user-facing change?**:
```release-note
None
```

### CODE REVIEW

File: cmd/csi-node-driver-registrar/node_register.go [hunk 1]
```diff
@@ -74,6 +75,44 @@ func nodeRegister(
 		klog.Errorf("Registration Server stopped serving: %v", err)
 		os.Exit(1)
 	}
+
+	//Periodically check whether the socket exists
```
Comment (verult) [inline cmd/csi-node-driver-registrar/node_register.go]: `// and creates one if it doesn't`

File: cmd/csi-node-driver-registrar/node_register.go [hunk 2]
```diff
@@ -74,6 +75,44 @@ func nodeRegister(
...
+	for {
+		select {
+		case <-t.C:
+			_, err := os.Stat(socketPath)
+			if err != nil {
+				if !os.IsNotExist(err) {
+					klog.Errorf("failed to stat the socket %s with error: %+v", socketPath, err)
+					os.Exit(1)
+				} else {
+					_, err := os.Create(socketPath)
```
Comment (verult) [inline cmd/csi-node-driver-registrar/node_register.go]: Does the registration server need to be updated to listen to this socket?

File: cmd/csi-node-driver-registrar/node_register.go [hunk 3]
```diff
@@ -74,6 +75,44 @@ func nodeRegister(
 		klog.Errorf("Registration Server stopped serving: %v", err)
 		os.Exit(1)
 	}
+
+	//Periodically check whether the socket exists
+	go periodicallyCheck(socketPath)
```
[the remark this answers is not in the document]
Comment (zhucan) [reply cmd/csi-node-driver-registrar/node_register.go]: Whether we can add gcpcServer.Serve(lis) to the periodicallyCheck()？

File: cmd/csi-node-driver-registrar/node_register.go (hunk 1 above)
Comment (zhucan) [reply cmd/csi-node-driver-registrar/node_register.go]: Done

File: cmd/csi-node-driver-registrar/node_register.go [hunk 4]
```diff
@@ -74,6 +80,41 @@ func nodeRegister(
 		klog.Errorf("Registration Server stopped serving: %v", err)
 		os.Exit(1)
 	}
+
 	// If gRPC server is gracefully shutdown, exit
 	os.Exit(0)
 }
+
+func periodicallyCheck(socketPath string, registrar registerapi.RegistrationServer) {
```
Comment (verult) [inline cmd/csi-node-driver-registrar/node_register.go]: function name: what are you periodically checking?

File: cmd/csi-node-driver-registrar/node_register.go [hunk 5]
```diff
@@ -74,6 +80,41 @@ func nodeRegister(
...
+	for {
+		select {
+		case <-t.C:
+			_, err := os.Stat(socketPath)
+			if err != nil {
+				if !os.IsNotExist(err) {
+					klog.Errorf("failed to stat the socket %s with error: %+v", socketPath, err)
+					os.Exit(1)
+				} else {
+					var oldmask int
```
Comment (verult) [inline cmd/csi-node-driver-registrar/node_register.go]: This code block is duplicated. Can it be shared somehow?

File: cmd/csi-node-driver-registrar/node_register.go [hunk 6]
```diff
@@ -74,6 +80,41 @@ func nodeRegister(
...
+					if runtime.GOOS == "linux" {
+						// Default to only user accessible socket, caller can open up later if desired
+						oldmask, _ = umask(0077)
+					}
+
+					klog.Infof("Starting Registration Server at: %s\n", socketPath)
+					_, err := net.Listen("unix", socketPath)
```
Comment (verult) [inline cmd/csi-node-driver-registrar/node_register.go]: Does the listening object need to be passed to the gRPC server?

**Comment by zhucan** [comment]
> Thanks for the update! How are you verifying the changes work as intended?
@jingxu97 How can I enter the registrar container?

File: cmd/csi-node-driver-registrar/node_register.go (hunk 2 above)
Comment (zhucan) [reply cmd/csi-node-driver-registrar/node_register.go]: if  the reg-socket had been deleted,  the socket was always used？It's no necessary to update to listen to this socket. Only create it.

File: cmd/csi-node-driver-registrar/node_register.go (hunk 4 above)
Comment (zhucan) [reply cmd/csi-node-driver-registrar/node_register.go]: is there a suitable name for the function? @jingxu97

**Comment by zhucan** [comment]
> Thanks for the update! How are you verifying the changes work as intended?

I make the image  from  centos7  and when the pod is running, I try to delete the socket on the host in "/var/lib/kubelet/plugins_registry/". It works.

**Comment by zhucan** [comment]
@verult  At first time, When creating the new pod, there are some warnings:
image

File: cmd/csi-node-driver-registrar/node_register.go [hunk 7]
```diff
@@ -36,6 +37,15 @@ func nodeRegister(
 	// pluginswatcher infrastructure. Node labeling is done by kubelet's csi code.
 	registrar := newRegistrationServer(csiDriverName, *kubeletRegistrationPath, supportedVersions)
 	socketPath := fmt.Sprintf("/registration/%s-reg.sock", csiDriverName)
+
+	ticker := time.NewTicker(5 * time.Second)
+	for {
+		<-ticker.C
+		go periodicallyCheck(socketPath, registrar)
```
Comment (verult) [inline cmd/csi-node-driver-registrar/node_register.go]: Do we need to spawn a new goroutine? Is there ever a case where reconciling the socket could take a long time?

File: cmd/csi-node-driver-registrar/node_register.go [hunk 8]
```diff
@@ -36,6 +37,15 @@ func nodeRegister(
...
+	}
+}
+
+func periodicallyCheck(socketPath string, registrar registerapi.RegistrationServer) {
```
Comment (verult) [inline cmd/csi-node-driver-registrar/node_register.go]: This function can be called something like "reconcileSocket", to be more descriptive of what it does.

**Comment by davidz627** [comment]
@zhucan is this PR still active?

**Comment by zhucan** [comment]
> @zhucan is this PR still active?

@davidz627 I‘m sorry, I don't have time to finish it now.

**Pull request closed without merging**